# Final Project-2: Credit Card Fraud Detection Model using Logistic Regression

## Candidate Name: Mayur Kumar Sharma
## Batch: DW-32

### 1. Importing dependencies

In [34]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report

### 2. Importing dataset, checking information, shape & Null-values

In [2]:
credit_card_data=pd.read_csv('creditcard.csv')
credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [4]:
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [3]:
credit_card_data.shape

(284807, 31)

In [5]:
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

### 3. Distribution of legit & fraud transactions

In [7]:
credit_card_data['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

#### 0-----> Legit
#### 1-----> Fraud

### 4. Balanced or Unbalanced check

In [10]:
credit_card_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

In [8]:
legit = credit_card_data[credit_card_data['Class']==0]
fraud = credit_card_data[credit_card_data['Class']==1]

print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


#### The distribution shows that this dataset is highly un-balanced

In [11]:
credit_card_data.describe()

Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  1.168375e-15  3.416908e-16 -1.379537e-15  2.074095e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     54201.500000 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%     84692.000000  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%    139320.500000  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   9.604066e-16  1.487313e-15 -5.556467e-16  1.213481e-16 -2.406331e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -6.915971e-01 -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01   
50%   -5.433583e-02 -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02   
75%    6.119264e-01  3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.654067e-16 -3.568593e-16  2.578648e-16  4.473266e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%    ... -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%    ...  1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   5.340915e-16  1.683437e-15 -3.660091e-16 -1.227390e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       5.600000   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02      22.000000   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02      77.165000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

In [12]:
legit.describe()

Time             V1             V2             V3  \
count  284315.000000  284315.000000  284315.000000  284315.000000   
mean    94838.202258       0.008258      -0.006271       0.012171   
std     47484.015786       1.929814       1.636146       1.459429   
min         0.000000     -56.407510     -72.715728     -48.325589   
25%     54230.000000      -0.917544      -0.599473      -0.884541   
50%     84711.000000       0.020023       0.064070       0.182158   
75%    139333.000000       1.316218       0.800446       1.028372   
max    172792.000000       2.454930      18.902453       9.382558   

                  V4             V5             V6             V7  \
count  284315.000000  284315.000000  284315.000000  284315.000000   
mean       -0.007860       0.005453       0.002419       0.009637   
std         1.399333       1.356952       1.329913       1.178812   
min        -5.683171    -113.743307     -26.160506     -31.764946   
25%        -0.850077      -0.689398      -0.766847      -0.551442   
50%        -0.022405      -0.053457      -0.273123       0.041138   
75%         0.737624       0.612181       0.399619       0.571019   
max        16.875344      34.801666      73.301626     120.589494   

                  V8             V9  ...            V21            V22  \
count  284315.000000  284315.000000  ...  284315.000000  284315.000000   
mean       -0.000987       0.004467  ...      -0.001235      -0.000024   
std         1.161283       1.089372  ...       0.716743       0.723668   
min       -73.216718      -6.290730  ...     -34.830382     -10.933144   
25%        -0.208633      -0.640412  ...      -0.228509      -0.542403   
50%         0.022041      -0.049964  ...      -0.029821       0.006736   
75%         0.326200       0.598230  ...       0.185626       0.528407   
max        18.709255      15.594995  ...      22.614889      10.503090   

                 V23            V24            V25            V26  \
count  284315.000000  284315.000000  284315.000000  284315.000000   
mean        0.000070       0.000182      -0.000072      -0.000089   
std         0.621541       0.605776       0.520673       0.482241   
min       -44.807735      -2.836627     -10.295397      -2.604551   
25%        -0.161702      -0.354425      -0.317145      -0.327074   
50%        -0.011147       0.041082       0.016417      -0.052227   
75%         0.147522       0.439869       0.350594       0.240671   
max        22.528412       4.584549       7.519589       3.517346   

                 V27            V28         Amount     Class  
count  284315.000000  284315.000000  284315.000000  284315.0  
mean       -0.000295      -0.000131      88.291022       0.0  
std         0.399847       0.329570     250.105092       0.0  
min       -22.565679     -15.430084       0.000000       0.0  
25%        -0.070852      -0.052950       5.650000       0.0  
50%         0.001230       0.011199      22.000000       0.0  
75%         0.090573       0.077962      77.050000       0.0  
max        31.612198      33.847808   25691.160000       0.0  

[8 rows x 31 columns]

In [13]:
fraud.describe()

Time          V1          V2          V3          V4  \
count     492.000000  492.000000  492.000000  492.000000  492.000000   
mean    80746.806911   -4.771948    3.623778   -7.033281    4.542029   
std     47835.365138    6.783687    4.291216    7.110937    2.873318   
min       406.000000  -30.552380   -8.402154  -31.103685   -1.313275   
25%     41241.500000   -6.036063    1.188226   -8.643489    2.373050   
50%     75568.500000   -2.342497    2.717869   -5.075257    4.177147   
75%    128483.000000   -0.419200    4.971257   -2.276185    6.348729   
max    170348.000000    2.132386   22.057729    2.250210   12.114672   

               V5          V6          V7          V8          V9  ...  \
count  492.000000  492.000000  492.000000  492.000000  492.000000  ...   
mean    -3.151225   -1.397737   -5.568731    0.570636   -2.581123  ...   
std      5.372468    1.858124    7.206773    6.797831    2.500896  ...   
min    -22.105532   -6.406267  -43.557242  -41.044261  -13.434066  ...   
25%     -4.792835   -2.501511   -7.965295   -0.195336   -3.872383  ...   
50%     -1.522962   -1.424616   -3.034402    0.621508   -2.208768  ...   
75%      0.214562   -0.413216   -0.945954    1.764879   -0.787850  ...   
max     11.095089    6.474115    5.802537   20.007208    3.353525  ...   

              V21         V22         V23         V24         V25         V26  \
count  492.000000  492.000000  492.000000  492.000000  492.000000  492.000000   
mean     0.713588    0.014049   -0.040308   -0.105130    0.041449    0.051648   
std      3.869304    1.494602    1.579642    0.515577    0.797205    0.471679   
min    -22.797604   -8.887017  -19.254328   -2.028024   -4.781606   -1.152671   
25%      0.041787   -0.533764   -0.342175   -0.436809   -0.314348   -0.259416   
50%      0.592146    0.048434   -0.073135   -0.060795    0.088371    0.004321   
75%      1.244611    0.617474    0.308378    0.285328    0.456515    0.396733   
max     27.202839    8.361985    5.466230    1.091435    2.208209    2.745261   

              V27         V28       Amount  Class  
count  492.000000  492.000000   492.000000  492.0  
mean     0.170575    0.075667   122.211321    1.0  
std      1.376766    0.547291   256.683288    0.0  
min     -7.263482   -1.869290     0.000000    1.0  
25%     -0.020025   -0.108868     1.000000    1.0  
50%      0.394926    0.146344     9.250000    1.0  
75%      0.826029    0.381152   105.890000    1.0  
max      3.052358    1.779364  2125.870000    1.0  

[8 rows x 31 columns]

### 5. Creating balanced dataset
#### Build a sample dataset containing similar distribution of Legit transactions and Fraud transactions (both equal to 492)

In [14]:
legit_sample = legit.sample(n=492)

In [15]:
new_dataset = pd.concat([legit_sample, fraud], axis=0)

In [16]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
141154   84152.0  1.082102  0.048270  0.502715  1.152298 -0.414798 -0.221686   
205124  135601.0  2.218649 -1.403706 -1.467122 -1.618902 -0.950464 -0.847996   
67148    52390.0  1.250676  0.274174  0.120656  0.994455  0.208206  0.098961   
129652   79172.0 -0.504884  0.777214  1.361708 -0.124435  0.261765 -0.324686   
47671    43303.0  1.164853  0.090559  0.248824  1.039241 -0.434522 -0.689717   

              V7        V8        V9  ...       V21       V22       V23  \
141154 -0.169016  0.198495  0.011677  ...  0.029020 -0.047005  0.028901   
205124 -0.704956 -0.240544 -1.660402  ... -0.284733 -0.485940  0.189870   
67148  -0.022691  0.017545  0.086815  ...  0.003787  0.097271 -0.145088   
129652  0.449755  0.185075 -0.150638  ... -0.231778 -0.669115 -0.000011   
47671  -0.016381  0.039729  0.068820  ...  0.063621  0.041042 -0.100092   

             V24       V25       V26       V27       V28  Amount  Class  
141154  0.163609  0.329808 -0.419355  0.016490  0.014079   27.50      0  
205124 -0.387034 -0.153038 -0.279606 -0.032069 -0.065155   59.95      0  
67148  -0.773023  0.647683 -0.246507  0.037741  0.011696    1.00      0  
129652 -0.344747 -0.262573  0.127990  0.268318  0.112986    2.58      0  
47671   0.279812  0.573418 -0.339959 -0.008526  0.008388   22.50      0  

[5 rows x 31 columns]

In [17]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [19]:
new_dataset['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

#### Now this "new_dataset" has balanced data.

In [18]:
new_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94815.558943 -0.092304 -0.043854 -0.004398 -0.019943  0.087599   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.030011 -0.074401  0.025315  0.062258  ... -0.010075 -0.006324   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.028330 -0.018003  0.057979 -0.010730 -0.007007  0.015291 -0.008489   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       76.292805  
1      122.211321  

[2 rows x 30 columns]

### 6. Splitting the dataset into Features (X) & Targets (Y)

In [20]:
X = new_dataset.drop(columns='Class', axis=1)
Y = new_dataset['Class']

In [21]:
X

Time        V1        V2        V3        V4        V5        V6  \
141154   84152.0  1.082102  0.048270  0.502715  1.152298 -0.414798 -0.221686   
205124  135601.0  2.218649 -1.403706 -1.467122 -1.618902 -0.950464 -0.847996   
67148    52390.0  1.250676  0.274174  0.120656  0.994455  0.208206  0.098961   
129652   79172.0 -0.504884  0.777214  1.361708 -0.124435  0.261765 -0.324686   
47671    43303.0  1.164853  0.090559  0.248824  1.039241 -0.434522 -0.689717   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V20       V21       V22  \
141154 -0.169016  0.198495  0.011677  ... -0.247749  0.029020 -0.047005   
205124 -0.704956 -0.240544 -1.660402  ... -0.390654 -0.284733 -0.485940   
67148  -0.022691  0.017545  0.086815  ... -0.156082  0.003787  0.097271   
129652  0.449755  0.185075 -0.150638  ...  0.012790 -0.231778 -0.669115   
47671  -0.016381  0.039729  0.068820  ... -0.243080  0.063621  0.041042   
...          ...       ...       ...  ...       ...       ...       ...   
279863 -0.882850  0.697211 -2.064945  ...  1.252967  0.778584 -0.319189   
280143 -1.413170  0.248525 -1.127396  ...  0.226138  0.370612  0.028234   
280149 -2.234739  1.210158 -0.652250  ...  0.247968  0.751826  0.834108   
281144 -2.208002  1.058733 -1.632333  ...  0.306271  0.583276 -0.269209   
281674  0.223050 -0.068384  0.577829  ... -0.017652 -0.164350 -0.295135   

             V23       V24       V25       V26       V27       V28  Amount  
141154  0.028901  0.163609  0.329808 -0.419355  0.016490  0.014079   27.50  
205124  0.189870 -0.387034 -0.153038 -0.279606 -0.032069 -0.065155   59.95  
67148  -0.145088 -0.773023  0.647683 -0.246507  0.037741  0.011696    1.00  
129652 -0.000011 -0.344747 -0.262573  0.127990  0.268318  0.112986    2.58  
47671  -0.100092  0.279812  0.573418 -0.339959 -0.008526  0.008388   22.50  
...          ...       ...       ...       ...       ...       ...     ...  
279863  0.639419 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00  
280143 -0.145640 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76  
280149  0.190944  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89  
281144 -0.456108 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00  
281674 -0.072173 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53  

[984 rows x 30 columns]

In [23]:
Y

141154    0
205124    0
67148     0
129652    0
47671     0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64

### 7. Split the data into Training data & Testing Data

In [24]:
x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=2)

In [25]:
print(X.shape, x_train.shape, x_test.shape)

(984, 30) (787, 30) (197, 30)


In [26]:
print(Y.shape, y_train.shape, y_test.shape)

(984,) (787,) (197,)


### 8. Model training : Logistic Regression

In [29]:
model = LogisticRegression()

In [30]:
model.fit(x_train, y_train)

LogisticRegression()

### 9. Model evaluation on training data

In [32]:
x_train_predicted = model.predict(x_train)
train_accuracy = accuracy_score(x_train_predicted, y_train)
train_accuracy

0.9440914866581956

### 10. Model evaluation on test data

In [33]:
x_test_predicted = model.predict(x_test)
test_accuracy = accuracy_score(x_test_predicted, y_test)
test_accuracy

0.949238578680203

### 11. Precision, Recall and F1 score

In [36]:
report = classification_report(x_test_predicted, y_test)
print(report)

              precision    recall  f1-score   support

           0       0.99      0.92      0.95       107
           1       0.91      0.99      0.95        90

    accuracy                           0.95       197
   macro avg       0.95      0.95      0.95       197
weighted avg       0.95      0.95      0.95       197



### 12. Results/Outcomes

#### (a). After successfully training the model with Logistic Regression Algo, we applied the testing for legit or fraud transaction.
#### (b). By comparing the model predictions on test data and actual result-data from dataset, we conclude that our model has accuracy of 95%. 
#### (c). The total applied test-data = 197
#### (d). Predicted Legit = 107 with accuracy of 95%
#### (e). Predicted Fraud = 90 with accuracy of 95%